### PR TITLE
GH-2485 fix failing test in RDFCollections

### PR DESCRIPTION
**Pull request by jeenbroekstra**
GitHub issue resolved: #2485 

Briefly describe the changes proposed in this PR:

- make sure test for cycle detection uses a collection that is well-formed (apart from the cycle obviously)
---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by jeenbroekstra** [comment]
Ok. So it looks as if the guava version bump was not directly responsible for this problem: even when going back to the old version, we still have a failing test.

**Comment by jeenbroekstra** [comment]
I'm going to take a closer look at the test itself.

**Comment by jeenbroekstra** [comment]
🤨

**Comment by hmottestad** [comment]
So you think it's the test that's actually broken, and not the underlying code?

**Comment by jeenbroekstra** [comment]
I don't know what to think. I reorganised the tests not because I thought that would fix it, but to isolate the failing function (cycle detection).

Then again I can't spot anything in the code that would make this test fail in some circumstances but not others. There's no IO involved, no concurrency, so no possible race conditions that I can see.

**Comment by hmottestad** [comment]
I found out the issue only occurs when the head node is lexicographically after the other nodes in the rdf list. So I pushed a commit where the head node is "z", which on my computer is always after since the other blank nodes all start with node... .

**Comment by jeenbroekstra** [comment]
> I found out the issue only occurs when the head node is lexicographically after the other nodes in the rdf list. So I pushed a commit where the head node is "z", which on my computer is always after since the other blank nodes all start with node... .

That is an interesting find, thanks! Gives us a good lead (though why lexicographic order of bnods should play a role in this is currently a mystery to me - perhaps a hash issue in the specific Model implementation we use?)

**Comment by jeenbroekstra** [comment]
Part of the problem is that the test introduces a cycle by adding a _second_ `rdf:rest` statement for the head node. The original `rdf:rest` statement is not replaced. If the head node is lexicographically _behind_ the object node of that original `rdf:rest` statement, the algorithm just ignores the second statement.

**Comment by jeenbroekstra** [comment]
Since the utility function is documented as expecting a well-formed collection, I've fixed this in the short term as a problem in the test, rather than in the code itself. The test now replaces the `rdf:rest` relation, rather than adding a second one, so that (apart from the introduced cycle) the collection is well-formed.

**Comment by jeenbroekstra** [comment]
<insert noughty word>! I _really_ hate it when the build fails on failing to download a dependency... multiple times.

**Comment by jeenbroekstra** [comment]
Alright, build is green, I'd appreciate a final review (and if possible, a green tick :)

File: core/model/src/test/java/org/eclipse/rdf4j/model/util/RDFCollectionsTest.java [hunk 1]
```diff
@@ -71,35 +70,33 @@ public void testConversionRoundtrip() {
...
-		}
+		RDFCollections.asValues(m, head, new ArrayList<>());
+	}
+
+	@Test(expected = ModelException.class)
+	public void testNonWellformedCollection_IncorrectHeadNode() {
+		Resource head = vf.createBNode();
+		Model m = RDFCollections.asRDF(values, head, new TreeModel());
+
+		RDFCollections.asValues(m, vf.createBNode(), new ArrayList<>());
```
Comment (hmottestad) [inline core/model/src/test/java/org/eclipse/rdf4j/model/util/RDFCollectionsTest.java:99]: Could you add the comment: `// supply incorrect head node` 

Makes the test easier to read.

**Review by hmottestad — APPROVED** [review]
Looks good to me. Just added one comment. 👍

File: core/model/src/test/java/org/eclipse/rdf4j/model/util/RDFCollectionsTest.java (hunk 1 above)
Comment (jeenbroekstra) [reply core/model/src/test/java/org/eclipse/rdf4j/model/util/RDFCollectionsTest.java:100]: Fixed.

**Pull request merged**